app/services/action_builder.rb

```diff
@@ -20,6 +20,15 @@ def build_action(extra_attrs = {})
   def previous_action
```

Comment:
Not exactly sure about the model and how page/campaign/action relate. Should ```page``` be given preference over the pages associated through the campaign (which includes ```page``` but perhaps not on the first index?). For the _join_ variant: you can use ```find_by```instead of ```where / first```